[flaky-tests] fix testSend and testRecv in ProcessGroupGlooTest

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #32242 [flaky-tests] skip testExceptions in ProcessGroupGloo if built with TSAN
* #32138 use gtest asserts in ProcessGroupGlooTest instead of other checks
* **#32134 [flaky-tests] fix testSend and testRecv in ProcessGroupGlooTest**

These tests weren't written in the most correct way and were often
flaky. It was tricky to identify these tests as flaky until we moved this file
to use gtest.

The gist of the issue is that the test previously would not coordinate sends
and recvs properly. For example, we created a single thread to test an
abortRecv and a successful recv. A separate sender thread was used to send 2
messages. What could go wrong here is that the first send could successfully
complete, resulting in the receiving end processing the message before it gets
the abort signal. In this case we would have an error in the test.

Differential Revision: [D19379395](https://our.internmc.facebook.com/intern/diff/D19379395/)

rohan-varma committed:
[flaky-tests] fix testSend and testRecv in ProcessGroupGlooTest

These tests weren't written in the most correct way and were often
flaky. It was tricky to identify these tests as flaky until we moved this file
to use gtest.

The gist of the issue is that the test previously would not coordinate sends
and recvs properly. For example, we created a single thread to test an
abortRecv and a successful recv. A separate sender thread was used to send 2
messages. What could go wrong here is that the first send could successfully
complete, resulting in the receiving end processing the message before it gets
the abort signal. In this case we would have an error in the test.

Differential Revision: [D19379395](https://our.internmc.facebook.com/intern/diff/D19379395/)

[ghstack-poisoned]

kostmo commented:
## :pill: CircleCI build failures summary and remediations
As of commit 92a4a5a3:


* **5/5** failures introduced in this PR


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=92a4a5a3c02531847ff7f8e5f4f37b9ee5d632bb).

### :detective: 2 new failures recognized by patterns
The following build failures do not appear to be due to upstream breakage:
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4225404) pytorch_xla_linux_xenial_py3_6_clang7_build (1/2)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=84470999) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=84470999))
<details>
<summary>
<code>Jan 13 22:16:11 RuntimeError: Failed to apply patch: /var/lib/jenkins/workspace/xla/torch_patches/X10-clip_grad.diff</code>
</summary>

```
Jan 13 22:16:11 + XDIR=/var/lib/jenkins/workspace/xla/scripts/.. 
Jan 13 22:16:11 + PTDIR=/var/lib/jenkins/workspace/xla/scripts/../.. 
Jan 13 22:16:11 + python /var/lib/jenkins/workspace/xla/scripts/cond_patch.py /var/lib/jenkins/workspace/xla/scripts/../torch_patches /var/lib/jenkins/workspace/xla/scripts/../.. 
Jan 13 22:16:11 1 out of 1 hunk FAILED 
Jan 13 22:16:11 Applying patch file: /var/lib/jenkins/workspace/xla/torch_patches/X10-clip_grad.diff 
Jan 13 22:16:11 Traceback (most recent call last): 
Jan 13 22:16:11   File "/var/lib/jenkins/workspace/xla/scripts/cond_patch.py", line 67, in <module> 
Jan 13 22:16:11     patch_repo(args) 
Jan 13 22:16:11   File "/var/lib/jenkins/workspace/xla/scripts/cond_patch.py", line 49, in patch_repo 
Jan 13 22:16:11     raise RuntimeError('Failed to apply patch: {}'.format(ppath)) 
Jan 13 22:16:11 RuntimeError: Failed to apply patch: /var/lib/jenkins/workspace/xla/torch_patches/X10-clip_grad.diff 
Jan 13 22:16:11 + cleanup 
Jan 13 22:16:11 + retcode=1 
Jan 13 22:16:11 + set +x 
Jan 13 22:16:11 =================== sccache compilation log =================== 
Jan 13 22:16:11 =========== If your build fails, please take a look at the log above for possible reasons =========== 
Jan 13 22:16:11 Compile requests                 1 
Jan 13 22:16:11 Compile requests executed        0 
Jan 13 22:16:11 Cache hits                       0 
Jan 13 22:16:11 Cache misses                     0 
Jan 13 22:16:11 Cache timeouts                   0 
```

</details>

#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4225412) caffe2_onnx_py3_6_clang7_ubuntu16_04_build (2/2)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=84476904) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=84476904))
<details>
<summary>
<code>Jan 13 22:18:17 CMakeFiles/ProcessGroupGlooTest.dir/ProcessGroupGlooTest.cpp.o:(.data.rel.ro+0x230): undefined reference to `typeinfo for testing::Test'</code>
</summary>

```
Jan 13 22:18:17 CMakeFiles/ProcessGroupGlooTest.dir/ProcessGroupGlooTest.cpp.o:(.data.rel.ro+0x138): undefined reference to `testing::Test::SetUp()' 
Jan 13 22:18:17 CMakeFiles/ProcessGroupGlooTest.dir/ProcessGroupGlooTest.cpp.o:(.data.rel.ro+0x140): undefined reference to `testing::Test::TearDown()' 
Jan 13 22:18:17 CMakeFiles/ProcessGroupGlooTest.dir/ProcessGroupGlooTest.cpp.o:(.data.rel.ro+0x170): undefined reference to `typeinfo for testing::Test' 
Jan 13 22:18:17 CMakeFiles/ProcessGroupGlooTest.dir/ProcessGroupGlooTest.cpp.o:(.data.rel.ro+0x188): undefined reference to `testing::Test::~Test()' 
Jan 13 22:18:17 CMakeFiles/ProcessGroupGlooTest.dir/ProcessGroupGlooTest.cpp.o:(.data.rel.ro+0x198): undefined reference to `testing::Test::SetUp()' 
Jan 13 22:18:17 CMakeFiles/ProcessGroupGlooTest.dir/ProcessGroupGlooTest.cpp.o:(.data.rel.ro+0x1a0): undefined reference to `testing::Test::TearDown()' 
Jan 13 22:18:17 CMakeFiles/ProcessGroupGlooTest.dir/ProcessGroupGlooTest.cpp.o:(.data.rel.ro+0x1d0): undefined reference to `typeinfo for testing::Test' 
Jan 13 22:18:17 CMakeFiles/ProcessGroupGlooTest.dir/ProcessGroupGlooTest.cpp.o:(.data.rel.ro+0x1e8): undefined reference to `testing::Test::~Test()' 
Jan 13 22:18:17 CMakeFiles/ProcessGroupGlooTest.dir/ProcessGroupGlooTest.cpp.o:(.data.rel.ro+0x1f8): undefined reference to `testing::Test::SetUp()' 
Jan 13 22:18:17 CMakeFiles/ProcessGroupGlooTest.dir/ProcessGroupGlooTest.cpp.o:(.data.rel.ro+0x200): undefined reference to `testing::Test::TearDown()' 
Jan 13 22:18:17 CMakeFiles/ProcessGroupGlooTest.dir/ProcessGroupGlooTest.cpp.o:(.data.rel.ro+0x230): undefined reference to `typeinfo for testing::Test' 
Jan 13 22:18:17 clang: error: linker command failed with exit code 1 (use -v to see invocation) 
Jan 13 22:18:17 caffe2/lib_c10d/test/CMakeFiles/ProcessGroupGlooTest.dir/build.make:104: recipe for target 'bin/ProcessGroupGlooTest' failed 
Jan 13 22:18:17 make[2]: *** [bin/ProcessGroupGlooTest] Error 1 
Jan 13 22:18:17 CMakeFiles/Makefile2:11668: recipe for target 'caffe2/lib_c10d/test/CMakeFiles/ProcessGroupGlooTest.dir/all' failed 
Jan 13 22:18:17 make[1]: *** [caffe2/lib_c10d/test/CMakeFiles/ProcessGroupGlooTest.dir/all] Error 2 
Jan 13 22:18:17 make[1]: *** Waiting for unfinished jobs.... 
Jan 13 22:18:17 [ 95%] Building CXX object caffe2/torch/CMakeFiles/torch_python.dir/csrc/autograd/python_function.cpp.o 
Jan 13 22:18:17 [ 95%] Building CXX object caffe2/torch/CMakeFiles/torch_python.dir/csrc/autograd/python_hook.cpp.o 
Jan 13 22:18:18 [ 95%] Built target run_plan 
Jan 13 22:18:18 [ 95%] Building CXX object caffe2/torch/CMakeFiles/torch_python.dir/csrc/autograd/python_legacy_variable.cpp.o 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

rohan-varma committed:
Update on "[flaky-tests] fix testSend and testRecv in ProcessGroupGlooTest"


These tests weren't written in the most correct way and were often
flaky. It was tricky to identify these tests as flaky until we moved this file
to use gtest.

The gist of the issue is that the test previously would not coordinate sends
and recvs properly. For example, we created a single thread to test an
abortRecv and a successful recv. A separate sender thread was used to send 2
messages. What could go wrong here is that the first send could successfully
complete, resulting in the receiving end processing the message before it gets
the abort signal. In this case we would have an error in the test.

Differential Revision: [D19379395](https://our.internmc.facebook.com/intern/diff/D19379395/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[flaky-tests] fix testSend and testRecv in ProcessGroupGlooTest"


These tests weren't written in the most correct way and were often
flaky. It was tricky to identify these tests as flaky until we moved this file
to use gtest.

The gist of the issue is that the test previously would not coordinate sends
and recvs properly. For example, we created a single thread to test an
abortRecv and a successful recv. A separate sender thread was used to send 2
messages. What could go wrong here is that the first send could successfully
complete, resulting in the receiving end processing the message before it gets
the abort signal. In this case we would have an error in the test.

Differential Revision: [D19379395](https://our.internmc.facebook.com/intern/diff/D19379395/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[flaky-tests] fix testSend and testRecv in ProcessGroupGlooTest"


These tests weren't written in the most correct way and were often
flaky. It was tricky to identify these tests as flaky until we moved this file
to use gtest.

The gist of the issue is that the test previously would not coordinate sends
and recvs properly. For example, we created a single thread to test an
abortRecv and a successful recv. A separate sender thread was used to send 2
messages. What could go wrong here is that the first send could successfully
complete, resulting in the receiving end processing the message before it gets
the abort signal. In this case we would have an error in the test.

Differential Revision: [D19379395](https://our.internmc.facebook.com/intern/diff/D19379395/)

[ghstack-poisoned]

pritamdamania87 commented on torch/lib/c10d/test/ProcessGroupGlooTest.cpp:
```diff
@@ -268,87 +269,67 @@ void testSend(const std::string& path) {
+  waitSendThreadAbort.join();
+  ASSERT_FALSE(sendCompleted);
+
+  // Now create a separate sender thread to ensure that future waitsends can
```
nit: separate receiver thread or does this comment have to be below when we create a sendThread?

rohan-varma replied:
Right, it should be below

pritamdamania87 commented on torch/lib/c10d/test/ProcessGroupGlooTest.cpp:
```diff
@@ -268,87 +269,67 @@ void testSend(const std::string& path) {
-  senderThread.join();
-  receiverThread.join();
+  // Sender thread
+  std::thread sendThread([&]() { sendCompleted = sendWork->wait(); });
```
How does this work? Didn't we abort `sendWork` on line 282? Don't we need to create a new `sendWork` object here?

rohan-varma replied:
@pritamdamania87 If you look at the implementation of `ProcessGroup.cpp`, it is actually the underlying gloo `buffer_` that can be aborted (see the tests in https://github.com/facebookincubator/gloo/pull/232/files which this is designed to mirror). If we create a new `sendWork` here via `pg.send()`, it will create a new buffer internally in `ProcessGroup`, but if we have a successful receive, it will be processed by the previous `sendWork` object (i.e. it's underlying `buffer_`). So if we created a new `sendWork` here the old one will get a completion and this one won't, so the test will fail. If we want to change this behavior so that we do need to create a new `sendWork` object, then that'll require changes to `ProcessGroup` and/or gloo's `unbound_buffer`. In terms of implementation, I'm not sure which is more correct

mrshenli replied:
This behavior does look a little weird. It means that even if the sender has aborted the send, and then come back to check the result of `wait()`, it could still return `True`? This is `ProcessGroupGloo::SendWork`, right? Is it possible to destroy the `buffer_` when abort is called? Will that prevent this from happening?

rohan-varma replied:
> This behavior does look a little weird. It means that even if the sender has aborted the send, and then come back to check the result of `wait()`, it could still return `True`? 

Right, this is possible. It's not really "aborting the send" though, it's more like we aborted waiting for the send (write) to be completed. This doesn't entail that we're cancelling the send, we're just no longer blocking waiting for it. We've already scheduled the write (asynchronously or not). It could still be completed.

Likewise for the `recv` side, we've only aborted blocking on waiting for the receive to be completed, we can't really (it seems to me) cancel receiving the data itself (since we've already called `buf->recv()`. 

> This is `ProcessGroupGloo::SendWork`, right? Is it possible to destroy the `buffer_` when abort is called? Will that prevent this from happening?

I can look into this.

mrshenli replied:
Thanks. This shouldn't block us from landing this PR. This fix is great, and `abort` is not used by any applications other than us I think, because below is how we expose abort currently (will remove it):

```c++
          .def(
              "abort",
              &::c10d::ProcessGroup::barrier,
              py::arg("opts") = ::c10d::BarrierOptions(),
              py::call_guard<py::gil_scoped_release>())
```

For now, maybe we can add some comment to the `abort()` API describing the current behavior. The fix for destroying buffer can come in in followup PRs.

rohan-varma replied:
@mrshenli I'm pretty sure that this is a different abort function, the one that this tests is the `ProcessGroup::Work::abort()` function (https://github.com/pytorch/pytorch/blob/master/torch/lib/c10d/ProcessGroup.cpp#L47 , implemented by sublcasses `SendWork` and `RecvWork`) (which is also only used by us in ProcessGroupAgent), the one that you referred to seems to be calling `Processgroup::barrier()`.

pritamdamania87 commented on torch/lib/c10d/test/ProcessGroupGlooTest.cpp:
```diff
@@ -358,17 +339,14 @@ void testRecv(const std::string& path) {
-  receiverThread.join();
+  // Receiver thread.
+  std::thread receiverThread([&]() { recvCompleted = recvWork->wait(); });
   senderThread.join();
```
Same comments apply here as I mentioned for testSend above.

jjlilley commented on torch/lib/c10d/test/ProcessGroupGlooTest.cpp:
```diff
@@ -268,87 +269,67 @@ void testSend(const std::string& path) {
-
-  // Helper receiver thread to simulate a real send/recv pair.
-  std::thread receiverThread([&]() {
+  // test that waiting for work to be sent can be aborted successfully.
```
nit: much of the time, abort() may be called prior to wait() below, depending on OS/etc

If you want, there are ways to increase the probability of wait() being called before abort(), e.g. the equivalent of folly::Baton<> to wait for the thread body to be reached prior to calling abort()

rohan-varma replied:
In terms of correctness, this shouldn't be an issue, as even if we call `abort()` before `wait()` the wait should still be successfully interrupted. Although if we want to try to test the wait() before abort() sequence, then we could go with a baton-based approach.

mrshenli commented on torch/lib/c10d/test/ProcessGroupGlooTest.cpp:
```diff
@@ -268,87 +269,67 @@ void testSend(const std::string& path) {
-    work = pg.recv(tensors, srcRank, tag);
-    work->wait();
+    auto recvWork = pg.recv(tensors, srcRank, tag);
+    recvWork->wait();
```
For now, is this guaranteed to be successful because it missed the abort signal? In that case, shall we pass the result to the main thread and `ASSERT_TRUE` on it?

rohan-varma replied:
This test is supposed to be checking the status of `sendCompleted` since it is testing that waiting on `send` can be aborted, and it asserts on that variable

mrshenli reviewed (approved):
Thanks a lot for the fix! Fixing/changing the behavior of `abort` should not block this PR.

rohan-varma committed:
Update on "[flaky-tests] fix testSend and testRecv in ProcessGroupGlooTest"


These tests weren't written in the most correct way and were often
flaky. It was tricky to identify these tests as flaky until we moved this file
to use gtest.

The gist of the issue is that the test previously would not coordinate sends
and recvs properly. For example, we created a single thread to test an
abortRecv and a successful recv. A separate sender thread was used to send 2
messages. What could go wrong here is that the first send could successfully
complete, resulting in the receiving end processing the message before it gets
the abort signal. In this case we would have an error in the test.

Differential Revision: [D19379395](https://our.internmc.facebook.com/intern/diff/D19379395/)

[ghstack-poisoned]

rohan-varma commented:
All failures are unrelated, landing.

rohan-varma commented:
@mrshenli @pritamdamania87 Filed https://github.com/pytorch/pytorch/issues/32363 to ensure that we follow up on the discussion regarding what the intended API should look like.
